src/GraphQL/Execution/ExecutionHelper.cs

```diff
@@ -97,9 +97,16 @@ public static Variables GetVariableValues(Document document, ISchema schema, Var
                         Name = v.Name
                     };
 
-                    object variableValue = null;
-                    inputs?.TryGetValue(v.Name, out variableValue);
-                    variable.Value = GetVariableValue(document, schema, v, variableValue);
+                    if (inputs.TryGetValue(v.Name, out var variableValue))
+                    {
+                        variable.Value = GetVariableValue(document, schema, v, variableValue);
+                    }
+                    else
+                    {
+                        var value = GetVariableValue(document, schema, v, v.DefaultValue?.Value);
+                        if (value != null)
+                            variable.Value = value;
+                    }
```

Comment:
Checking for a default variable value has now been moved out of `GetVariableValue` into `GetVariableValues`.

Reply:
After looking at these methods, my intuition tells me that they should be rewritten in v4.